client/src/controllers/SwapController.ts

```diff
@@ -131,8 +149,13 @@ export class SwapController extends Controller<
     // set up initial loading state (if set originally in the HTML)
     this.loadingValue = false;
 
-    // set up debounced method
+    // set up debounced methods
+    this.replaceLazy = debounce(this.replace.bind(this), this.waitValue);
     this.searchLazy = debounce(this.search.bind(this), this.waitValue);
+    this.submitLazy = debounce(this.submit.bind(this), this.waitValue);
```

Comment:
Looks a bit funny but I think it's fine. Would be nice if we can do `anyMethod.debounce` everywhere but I know you said that it was tricky to implement.

Reply:
yep, I am happy to raise a new issue to look at our TypeScript version, I am confident-ish that once we adopt 5.0+ we can do some nicer things to abstract this away.

Alternatively, the next version of Stimulus should come with a way to make it easier to set up debouncing as an action option. Currently it's hard to pass variables or access the Controller instance from these methods - https://stimulus.hotwired.dev/reference/actions#options . However, in the next release my merged PRs in that repo make this possible.

- https://github.com/hotwired/stimulus/pull/692
- https://github.com/hotwired/stimulus/pull/691

This way we could avoid the debouncing set up / clean up in the controller methods and just write our actions like this.

```html
<div
  ...
  data-action="submit->w-swap#submit:stop:prevent input->w-swap#submit:debounce input->w-swap#submit:debounce navigate->w-swap#replace"
  data-w-swap-debounce-param="50" // 50ms, but default can still be set on the controller
  data-w-swap-debounce-value="50" // could be used also
>
```

We would track the function being debounced slightly differently but this will be much nicer and means we can actually remove various debounce/delay in other controller implementations and just write them as data-actions.